neurolib/models/multimodel/builder/adex.py

```diff
@@ -0,0 +1,944 @@
+        :type lin_nonlin_cascade_filename: str|None
+        :param seed: seed for random number generator
+        :type seed: int|None
+        """
+        params = self._rescale_strengths(params)
+        super().__init__(params=params, seed=seed)
+        # use the same file as neurolib's native
+        lin_nonlin_cascade_filename = lin_nonlin_cascade_filename or os.path.join(
+            os.path.dirname(os.path.realpath(__file__)), "..", "..", "aln", "aln-precalc", DEFAULT_CASCADE_FILENAME,
+        )
+        self._load_lin_nonlin_cascade(lin_nonlin_cascade_filename)
+
+    def _load_lin_nonlin_cascade(self, filename):
+        """
+        Load precomputed linear-nonlinear cascade from h5 file.
+        """
+        # load linear-nonlinear cascade quantities from file
+        logging.info(f"Reading precomputed quantities from {filename}")
+        loaded_h5 = File(filename, "r")
+        self.mu_range = np.array(loaded_h5["mu_vals"])
+        self.d_mu = self.mu_range[1] - self.mu_range[0]
+        self.sigma_range = np.array(loaded_h5["sigma_vals"])
+        self.d_sigma = self.sigma_range[1] - self.sigma_range[0]
+        self.firing_rate_cascade = np.array(loaded_h5["r_ss"])
+        self.voltage_cascade = np.array(loaded_h5["V_mean_ss"])
+        self.tau_cascade = np.array(loaded_h5["tau_mu_exp"])
+        logging.info("Loading done, all quantities loaded.")
+        # close the file
+        loaded_h5.close()
+        self.lin_nonlin_fname = filename
+
+    def update_params(self, params_dict):
+        """
+        Update parameters as in the base class but also rescale.
+        """
+        # if we are changing C_m or g_L, update tau_m as well
+        if any(k in params_dict for k in ("C_m", "g_L")):
+            C_m = params_dict["C_m"] if "C_m" in params_dict else self.params["C_m"]
+            g_L = params_dict["g_L"] if "g_L" in params_dict else self.params["g_L"]
+            params_dict["tau_m"] = C_m / g_L
+
+        # if we are chaning any of the J_exc_max, tau_syn_exc or c_global, rescale c_global
+        if any(k in params_dict for k in ("c_global", "J_exc_max", "tau_syn_exc")):
+            # get original c_global
+            c_global = (
+                params_dict["c_global"]
+                if "c_global" in params_dict
+                else self.params["c_global"] * (self.params["J_exc_max"] / self.params["tau_syn_exc"])
+            )
+            tau_syn_exc = params_dict["tau_syn_exc"] if "tau_syn_exc" in params_dict else self.params["tau_syn_exc"]
+            J_exc_max = params_dict["J_exc_max"] if "J_exc_max" in params_dict else self.params["J_exc_max"]
+            params_dict["c_global"] = c_global * tau_syn_exc / J_exc_max
+
+        # update all parameters finally
+        super().update_params(params_dict)
+
+    def describe(self):
+        return {
+            **super().describe(),
+            "lin_nonlin_cascade_filename": self.lin_nonlin_fname,
+        }
+
+    def _callbacks(self):
```

Comment:
now the famous callbacks:
for `jitcdde` backend it was tricky - it compiles symbolic code into C and then integrates - not doable with table lookup. the author of `jitcdde` was so nice, he implemented a functionality called `python_callbacks` (functions which you can call in derivatives definition)
the definition of callbacks is similar as for `sync` variables hence it must be a list of tuples, where each tuple is callback name (as `symengine.Symbol`) and then its definition as a python function, plus number of parameters (hence tuple of length 3).. of course, for speed callbacks are defined here as `numba` jitted functions